### ISSUE TITLE
optimizer regression for indirect queries on Lucene full-text and range indexes in eXist-2.2 and eXist-3.0.RC1

### ISSUE THREAD

**Issue by rvdb** [eXist-db/exist#873]
In eXist-2.2 and eXist-3.3.RC1, the optimizer seems to behave buggy for indirect queries on node sets that have been stored in a variable first. When such indirect queries are run on the Lucene full-text and range indexes, the search is either slow or incorrect, which suggests that the optimizer isn't working optimally.

For comparison, behaviour for both 'direct' and 'indirect' versions of queries will be described below, on both Lucene full-text and range indexes. These tests can be run straight on the eXist demo app.
- Lucene full-text index: indirect queries work, but take _much_ longer than direct queries
  - direct query: 419 results in under 0.020 sec.
    
    let $a := //SPEECH[ft:query(., 'hamlet')]
    return $a
  - indirect query: 419 results in **more than 2 sec.**
    
    let $a := //SPEECH
    return $a[ft:query(., 'hamlet')]
- Lucene range index: indirect queries don't return any results at all
  - direct query: 359 results in 0.018 sec.
    
    let $a := //SPEAKER[range:matches(., 'HAMLET')]
    return $a
  - indirect query: **0 results**
    
    let $a := //SPEAKER
    return $a[range:matches(., 'HAMLET')]

It seems that for indirect queries on the Lucene full-text and range indexes, the optimizer does not query the actual indexed nodes.

**Comment by adamretter** [comment]
I think this is one for @wolfgangmm as he has been looking at the Lucene index and regressions recently.

**Comment by adamretter** [comment]
* Copying from 
Ron van den Branden's post to the mailing list http://markmail.org/message/kmz6ecktup3ztvjr *

> I've just checked with the latest commit in the develop branch (d9ecd338b18eb1e6298881b5c2768c06b690a1ac), and the problem is still there.
> 
> Apart from my personal interest in this issue (I hope I'm not being too impatient) I wonder if this isn't an important issue in the light of a new release, since indirection seems quite a common coding pattern. For me, at least, this would be a major show-stopper for upgrading to eXist-3.0. Again, if there is anything I can do more besides reporting and testing (and fixing the issue myself, which I can't, unfortunately), I'd gladly do so. I hope the issue I reported at https://github.com/eXist-db/exist/issues/873 is clear and unambiguous; otherwise I would love to investigate how it can be improved.

**Comment by wolfgangmm** [comment]
My bugfix #949 addresses the failing range index queries. It does not improve the performance of indirect queries. This would be possible by making the optimizer inline the context variable (we do that for constants already). However, inlining is not always desirable. In some contexts inlining would have the contrary effect, in particular if there's no index defined. So the optimizer would need to take this into account. I'll have to think about this some more ...

**Comment by rvdb** [comment]
Ok, some progress, great! Is it wise, though, to close the issue? This way, you might risk sub-optimal performance for a common type of queries in eXist-3.0. Or should I isolate the performance of indirect queries in a separate issue?

**Comment by wolfgangmm** [comment]
Yes, right, the issue should not have been closed. Reopening it.

**Comment by Torcsi** [comment]
This issue is still there - it causes severe performance drop and blocks our upgrade to eXist 3.x.

A complete study is posted [here](http://www.arcdev.hu/temp/eXistDb_3_indexing_problem.pdf)
Rewriting our xqls is nearly impossible; the pinpointed case is just one from the many...

**Comment by rvdb** [comment]
I've checked again for a (simple) real-life query with eXist-3.6.0: indeed, a "direct" query expression against the Lucene full-text index takes under 0.5 seconds to complete, while reformulation as an "indirect" query expression (where the context node is first declared as a separate variable) explodes the query execution time to over 5 seconds. That's a factor 10, still consistent with my first observation, it seems.

**Comment by rvdb** [comment]
Ahh, fortunately I've found a workaround (might be of interest to @Torcsi, since it requires minimal XQuery rewriting): if the "indirect" variable is first intersected (or united) with itself (I hope the example will clarify), optimization seems to work again! 

Instead of the slow:

    let $a := //SPEECH
    return $a[ft:query(., 'hamlet')]

...which takes 2.74 seconds on the demo server to return 419 elements, this version does the job lightning fast:

    let $a := //SPEECH
    return ($a intersect $a)[ft:query(., 'hamlet')]

419 elements in 0.01 second, as you'd expect from eXist! The Optimizer works in mysterious ways...

**Comment by joewiz** [comment]
@rvdb That's crazy! Thanks for the tip. Mysterious indeed.

**Comment by wolfgangmm** [comment]
That's a weird tip! I really need to have a closer look at what the optimizer does here.

**Comment by Torcsi** [comment]
@rvdb Thank you, we are working to proof this in our working environment.

It should be carefully tested if the two above would not degrade.

Isn't this all related to some lazy evaluation of the query?

Optimizers **do** magic, so those writing optimizers are therefore magicians...

**Comment by Torcsi** [comment]
Tested and found "where" to be also removed/reorganized as described in the query optimization docs.

Speed is like in 2.2, so this issue does not blocking us. 
Other issues though still obstruct switch to 3.x

**Comment by Torcsi** [comment]
We have run into further issues and decided to make a complete test.
We have used 4.3 and our measures show 3.3-16x performance drop on synthetic exampples, totally obstructing porting to 4.3.
All former tips were tested.
The test environment and results are detailed [here](http://www.arcdev.hu/temp/eXistDb4.3performance/eXist4-3vs2-2.html)

Please advice what should we do. Is this the performance drop normal?

**Comment by duncdrum** [comment]
Fix + regression test in #6758, and a JMH benchmark (`LetBoundSourceLuceneBenchmark`, in `exist-indexes-jmh`) for the dashboard, tracking `let $a := //SPEECH return $a[ft:query(., ...)]` vs. the direct form.

**Root cause of the reported regression**, in two parts:
1. The parser wraps a FLWOR clause's terminal return expression (and an `if/then/else`'s else-branch) in a `DebuggableExpression`. That wrapper deliberately doesn't register itself as the AST parent when an ancestor already has one, so `Optimizer#visitFilteredExpr`'s rewrite resolves the filtered expression's parent to the `LetExpr` itself, not the wrapper -- and the resulting `replace()` call silently no-ops, since the `LetExpr`'s actual child slot holds the wrapper, not the filtered expression. The `(#exist:optimize#)` pragma got built but never wired into the executed tree.
2. Even once wired up, `Optimize#eval` discarded the index pre-selected candidate set whenever the filtered expression's source was a plain variable reference, because `VariableReference#eval` ignores whatever context it's given and always returns the variable's full, unfiltered value.

Both are fixed in #6758 for the exact shape reported here (`let $a := X return $a[pred]`, including the `outer//$a[pred]` abbreviated-step variant).

**Two related-but-distinct gaps found while building the benchmark, still open:**

- **`count($a[pred])` (or any typed function argument wrapping the filtered expression) is still slow.** `Function#checkArgument` wraps an already-analyzed argument in a `DynamicCardinalityCheck` (or `DynamicNameCheck`/`DynamicTypeCheck`/etc.) *after* `analyze()` has already fixed that argument's parent pointer -- so the same `replace()`-based rewrite can't find it there either, for the same underlying reason as (1) above, via a different wrapper class. `count(let $a := X return $a[pred])` (wrapping the whole `let`, not just the return) avoids it and is what the benchmark actually measures.
- **A second bracket-group predicate never gets optimized at all, independent of `let`.** `$a[ft:query(...)][SPEAKER = 'HAMLET']` parses as *nested* `FilteredExpression`s (one wrapping the other), and `FilteredExpression` itself doesn't implement `RewritableExpression`. The inner FE (holding the Optimizable predicate) can never get its own pragma, and the outer FE's own Optimizable-detection never looks inside the nested inner FE to find one. This predates the `let`-specific fix entirely and isn't specific to variables -- `doc('x')//a[ft:query(...)][@n=1]` would hit the same wall. Results stay correct, just unoptimized.

`LetBoundSourceLuceneBenchmark.shapeLetBoundSourceExtraPredicate` is kept in the benchmark specifically to track this second gap on the dashboard until it's closed. I'm looking into a fix for the typed-function-argument case next.
